### PR TITLE
[wip] phpunit and logfiles errorr

### PR DESCRIPTION
**Pull request by ddurieux**

### CODE REVIEW

File: inc/db.function.php [hunk 1]
```diff
@@ -1378,6 +1378,10 @@ function TableExists($tablename) {
 function FieldExists($table, $field, $usecache=true) {
    global $DB;
 
+   if (!TableExists($table)) {
```
Comment (trasher) [inline inc/db.function.php]: I guess we should log something here; it may be confusing

File: inc/db.function.php [hunk 2]
```diff
@@ -1399,6 +1403,10 @@ function FieldExists($table, $field, $usecache=true) {
 function isIndex($table, $field) {
    global $DB;
 
+   if (!TableExists($table)) {
```
Comment (trasher) [inline inc/db.function.php]: Idem

File: tests/bootstrap.php [hunk 3]
```diff
@@ -142,7 +142,12 @@ function loadDataset() {
    $CFG_GLPI['url_base']      = GLPI_URI;
    $CFG_GLPI['url_base_api']  = GLPI_URI . '/apirest.php';
 
-   @mkdir(GLPI_LOG_DIR, 0755, true);
+   if (!file_exists(GLPI_LOG_DIR)) {
```
Comment (remicollet) [inline tests/bootstrap.php]: This is not a file ;)

File: tests/bootstrap.php [hunk 4]
```diff
@@ -142,7 +142,12 @@ function loadDataset() {
...
+      @mkdir(GLPI_LOG_DIR, 0755, true);
```
Comment (remicollet) [inline tests/bootstrap.php]: If you add a test, drop the silent operator

File: tests/DbTestCase.php [hunk 5]
```diff
@@ -36,27 +36,33 @@ class DbTestCase extends PHPUnit_Framework_TestCase {
 
    protected function setUp() {
       global $DB;
+      
+      $this->Login();
```
Comment (remicollet) [inline tests/DbTestCase.php]: Login have a cost... not sure if it worth adding it to all tests

File: inc/networkport.class.php [hunk 6]
```diff
@@ -264,7 +264,8 @@ function splitInputForElements($input) {
       $this->getEmpty();
 
       foreach ($input as $field => $value) {
-         if (array_key_exists($field, $this->fields)) {
+         if (array_key_exists($field, $this->fields)
+                 || $field == '_no_history') {
```
Comment (remicollet) [inline inc/networkport.class.php]: Didn't we talk we should also remove the getEmpty call above (to not break history feature, diffing fields before / after ?)

Comment (remicollet) [inline inc/networkport.class.php]: Should be anough (better ?) to just skip all virtual fields

     $field[0] == '_'

Comment (remicollet) [inline inc/networkport.class.php]: And
```
$clone = clone $this;
$clone->getEmpty();
...
if (array_key_exists($field, $clone->fields)  || $field[0] == '_') {

```

File: tests/bootstrap.php (hunk 3 above)
Comment (ddurieux) [inline tests/bootstrap.php]: No but file_exist test file/directory cf http://php.net/manual/en/function.file-exists.php

File: tests/DbTestCase.php (hunk 5 above)
Comment (ddurieux) [inline tests/DbTestCase.php]: In test it have error:
```PHP Warning(2): session_regenerate_id(): Cannot regenerate session id - headers already sent```

File: inc/db.function.php (hunk 1 above)
Comment (remicollet) [inline inc/db.function.php]: raise a E_USER_ERROR, so will be logged in php_errors.

**Comment by remicollet** [comment]
I start thinking our error_handler is not very useful for unit test...

At least, having to dig in php-errors file in unit test, especially for non fatal or user error seems terrible bad.
Perhaps we should, (for unit test only) raise some exception... so we can easily detect then using expectedException / expectedExceptionMessage docblocks

**Comment by remicollet** [comment]
Tested (works for DBiterator test only, more work needed for other, login... already done in this PR):
```
diff --git a/inc/toolbox.class.php b/inc/toolbox.class.php
index f69712d..70f5090 100644
--- a/inc/toolbox.class.php
+++ b/inc/toolbox.class.php
@@ -572,6 +572,11 @@ class Toolbox {
       // Save error
       self::logInFile("php-errors", $err);
 
+      // For unit test
+      if (class_exists('GlpitestException')) {
+         throw new GlpitestException($err);
+      }
+
       return $errortype[$errno];
    }
 
diff --git a/tests/DBmysqlIteratorTest.php b/tests/DBmysqlIteratorTest.php
index 3b8cd9e..719e813 100644
--- a/tests/DBmysqlIteratorTest.php
+++ b/tests/DBmysqlIteratorTest.php
@@ -56,6 +56,10 @@ class DBmysqlIteratorTest extends DbTestCase {
    }
 
 
+   /**
+    * @expectedException        GlpitestException
+    * @expectedExceptionMessage Missing table name
+    */
    public function testNoTableWithWhere() {
 
       file_put_contents(GLPI_LOG_DIR . '/php-errors.log', '');
@@ -68,23 +72,27 @@ class DBmysqlIteratorTest extends DbTestCase {
    }
 
 
+   /**
+    * Temporarily, this is an error, will be allowed later
+    *
+    * @expectedException        GlpitestException
+    * @expectedExceptionMessage Missing table name
+    */
    public function testNoTableWithoutWhere() {
 
-      file_put_contents(GLPI_LOG_DIR . '/php-errors.log', '');
       $it = new DBmysqlIterator(NULL, '');
       $this->assertEquals('SELECT *', $it->getSql(), 'No table');
+   }
 
-      // Temporarily, this is an error, will be allowed later
-      $buf = file_get_contents(GLPI_LOG_DIR . '/php-errors.log');
-      $this->assertRegExp('/Missing table name/', $buf, 'Missing table');
 
-      file_put_contents(GLPI_LOG_DIR . '/php-errors.log', '');
+   /**
+    * @expectedException        GlpitestException
+    * @expectedExceptionMessage Missing table name
+    */
+   public function testNoTableWithoutWhereBis() {
+
       $it = new DBmysqlIterator(NULL, ['FROM' => []]);
       $this->assertEquals('SELECT *', $it->getSql(), 'No table');
-
-      // Temporarily, this is an error, will be allowed later
-      $buf = file_get_contents(GLPI_LOG_DIR . '/php-errors.log');
-      $this->assertRegExp('/Missing table name/', $buf, 'Missing table');
    }
 
 
diff --git a/tests/bootstrap.php b/tests/bootstrap.php
index c4dbf3f..d2d5a5a 100644
--- a/tests/bootstrap.php
+++ b/tests/bootstrap.php
@@ -41,6 +41,10 @@ include_once __DIR__ . '/DbTestCase.php';
 define('TU_USER', '_test_user');
 define('TU_PASS', 'PhpUnit_4');
 
+class GlpitestException extends Exception
+{
+}
+
 function loadDataset() {
    global $CFG_GLPI;
 
@@ -142,7 +146,7 @@ function loadDataset() {
    $CFG_GLPI['url_base']      = GLPI_URI;
    $CFG_GLPI['url_base_api']  = GLPI_URI . '/apirest.php';
 
-   @mkdir(GLPI_LOG_DIR, 0755, true);
+   is_dir(GLPI_LOG_DIR) or mkdir(GLPI_LOG_DIR, 0755, true);
 
    $conf = Config::getConfigurationValues('phpunit');
    if (isset($conf['dataset']) && $conf['dataset']==$data['_version']) {
```

**Comment by ddurieux** [comment]
Ok so like this we don;t need test data in php-errors.log (like I do in this PR), but test the exception

**Comment by remicollet** [comment]
@ddurieux yes, see https://github.com/glpi-project/glpi/pull/1167/commits/eac0884a663d50716d40609d44ae6b053aeeb2ae 

For now, only for E_ERROR, and E_USER_ERROR (so warnings or notices are still ignored)

**Comment by orthagh** [comment]
I archive this old PR, if you think about continue, please let me know, we will reopen.

**Pull request closed without merging**